File: src/main/java/org/pyjjs/scheduler/core/api/impl/actors/resource/placing/TimeLine.java
package org.pyjjs.scheduler.core.api.impl.actors.resource.placing;

import org.pyjjs.scheduler.core.api.impl.actors.resource.placing.algorithms.PlacingAlgorithm;

import javax.annotation.Nullable;
import java.util.*;

public class TimeLine {

    private Set<TimePart> freeTimes = new TreeSet<>((o1, o2) -> o1.start.compareTo(o2.start));
    private Set<UsedTime> usedTimes = new TreeSet<>((o1, o2) -> o1.timePart.start.compareTo(o2.timePart.start));

    private PlacingAlgorithm placingAlgorithm;
    private Long planningHorizon = Long.MAX_VALUE;

    public TimeLine() {
        freeTimes.add(new TimePart(0L));
    }

    public void setPlacingAlgorithm(PlacingAlgorithm placingAlgorithm) {
        this.placingAlgorithm = placingAlgorithm;
    }

    @Nullable
    public TimePart findNearestFreeTime() {
        Iterator<TimePart> freeTimeIterator = freeTimes.iterator();
        if(freeTimeIterator.hasNext()) {
            return freeTimeIterator.next();
        }
        return null;
    }

    @Nullable
    public UsedTime findNearestUsedTime() {
        Iterator<UsedTime> freeTimeIterator = usedTimes.iterator();
        if(freeTimeIterator.hasNext()) {
            return freeTimeIterator.next();
        }
        return null;
    }

    public class UsedTime {
        private TimePart timePart;
        private Object tenant;

        public UsedTime(Object tenant) {
            this.tenant = tenant;
            this.timePart = new TimePart();
        }

        public void setTimePart(TimePart timePart) {
            this.timePart = timePart;
        }

        public TimePart getTimePart() {
            return timePart;
        }

        public Object getTenant() {
            return tenant;
        }

        public void setTenant(Object tenant) {
            this.tenant = tenant;
        }
    }

    public class TimePart {
        private Long start = 0L;
        private Long duration = 0L;

        public TimePart() {
        }

        public TimePart(Long start) {
            this.start = start;
        }

        public Long getStart() {
            return start;
        }

        public void setStart(Long start) {
            this.start = start;
        }

        public Long getDuration() {
            return duration;
        }

        public void setDuration(Long duration) {
            this.duration = duration;
        }
    }
}

File: src/main/kotlin/org/pyjjs/scheduler/core/api/impl/actors/resource/placing/TimeLine.kt
package org.pyjjs.scheduler.core.api.impl.actors.resource.placing

import org.pyjjs.scheduler.core.api.impl.actors.resource.placing.algorithms.PlacingAlgorithm
import java.util.*

class TimeLine {

    private val freeTimes = TreeSet<TimePart> { o1, o2 -> o1.start.compareTo(o2.start) }
    private val usedTimes = TreeSet<UsedTime> { o1, o2 -> o1.timePart.start.compareTo(o2.timePart.start) }

    private var placingAlgorithm: PlacingAlgorithm? = null
    private val planningHorizon = java.lang.Long.MAX_VALUE

    init {
        freeTimes.add(TimePart(0L))
    }

    fun setPlacingAlgorithm(placingAlgorithm: PlacingAlgorithm) {
        this.placingAlgorithm = placingAlgorithm
    }

    fun findNearestFreeTime(): TimePart? {
        val freeTimeIterator = freeTimes.iterator()
        if (freeTimeIterator.hasNext()) {
            return freeTimeIterator.next()
        }
        return null
    }

    fun findNearestUsedTime(): UsedTime? {
        val freeTimeIterator = usedTimes.iterator()
        if (freeTimeIterator.hasNext()) {
            return freeTimeIterator.next()
        }
        return null
    }

    inner class UsedTime(var tenant: Any) {
        var timePart: TimePart = TimePart()
    }

    inner class TimePart {
        var start: Long = 0L
        var duration: Long = 0L

        constructor()

        constructor(start: Long) {
            this.start = start
        }
    }
}
